test/core/test_expression_constructor.py

```diff
@@ -1,38 +1,40 @@
-from mathics.core.expression import Expression
+from mathics.core.expression import Expression, to_expression, ElementsProperties
```

Comment:
@rocky, I started to wonder if `build_expression` or `create_expression` wouldn't be a better name for  `to_expression`: I used `to_expression` to denote a method that takes an object and converts it into an expression. On the other hand, the function `to_expression` is taking several elements and properties, and with them, making conversions and finally assembling an Expression. 
I ask about this now, because at some point we need to do a large number of changes over all the modules.

Reply:
`to_xxx` as a function takes  whatever you give and turns it into `xxx`. As top-level functions already have 

* `to_python`, 
* `to_sympy` 
* `to_real_or_none`
* `to_integer_or_none`
* `to_complex`
* `to_boxes`
* `to_sympy_matrix`
* `to_mpmath_matrix`

among others. 

The `or_none` in these I kind of think of as superfluous and the `or_none` should be removed, since `None` is the value that gets returned when a value doesn't fit and we don't want to throw an error. So this fits into the pattern way more nicely that `build` or `create` do.

As an instance method, as you noted there is there is a `to_expression()` instance method. Notice there are also `to_python()`, and `to_sympy()` _instance_ methods and again this fits into this pattern of having both instance methods and top_level methods that return the thing after the `to_` as an object. 

More generally in OO programs, a `to_xxx` pattern is what is generally what is used as a helper function to create an instance of some sort of special object.

So in sum, yes, there is a pattern and this is just continuing the pattern. I am sorry that this pattern was not familiar or noticed either in this code or in general. And that was a part of into why I wanted to start  a `ListExpression()` object even though right now it is woefully inadequate in terms of being an upward-compatible specialization of `Expression()` and needs a lot of work still such as to create a base class that both of these inherit from.

Know your strengths and also know your limitations. And although among others I would not consider myself having a flair for API design, among the people we have doing it here, it makes me seem more talented. I suppose a part of it is just that I've seen other interpreters and lots of other code Python and others.

Reply:
> `to_xxx` as a function takes whatever you give and turns it into `xxx`. As top-level functions already have
> 
>     * `to_python`,
> 
>     * `to_sympy`
> 
>     * `to_real_or_none`
> 
>     * `to_integer_or_none`
> 
>     * `to_complex`
> 
>     * `to_boxes`
> 
>     * `to_sympy_matrix`
> 
>     * `to_mpmath_matrix`
> 
> 
> among others.

OK, but in all the cases in the list, the method takes an object of a certain class and returns an expression. The new function, on the other hand, takes several elements and builds the expression. 

> Know your strengths and also know your limitations. And although among others I would not consider myself having a flair for API design, among the people we have doing it here, it makes me seem more talented. I suppose a part of it is just that I've seen other interpreters and lots of other code Python and others.


I guess that I didn't express myself well. Sorry for that. The aim was not to question your experience and knowledge. I asked you about this because I made myself one idea about what this pattern ("to_*" ) means, and the change you proposed seemed to me was in contradiction with that idea. So, I thought that I was missing something.  As I told you many times, I am OK with any name you choose, but I would like to understand the choice to avoid screwing up in future rounds.

Reply:
> 
> OK, but in all the cases in the list, the method takes an object of a certain class and returns an expression. The new function, on the other hand, takes several elements and builds the expression.
> 

I think this is a red herring. First, most of the conversion functions like `to_python` or `to_real`  input and output something that isn't considered a compound object. For the few that do, the places that want to use these happen to have the collections already stored in a single variable, so there isn't a need or convenience to break this out. You won't find many class constructors in Mathics that benefit by taking an arbitrary number of arguments and then need to bundle those as a collection. So the to_xxx helper method is just mirroring that aspect of the class.

The other place where it comes up so far and for the same reason is `to_mathics_list()` which is a construction helper for `ListExpression()`

An analogy would be like saying we can't create a ternary "op" functions because all of the other ops take one or two arguments.  

So in sum, the `*elements` is an aspect or convenience dictated by how you want to call the function. The name of a function generally isn't directed by the form you want to pass arguments to the function.